File: frontend/src/lib/cbsServiceTaxonomy.ts
/** CBS service taxonomy helpers (mirrors backend `cbs_service_taxonomy`). */

export const UNCLASSIFIED_DOMAIN = "UNCLASSIFIED";

const DOMAIN_BY_PREFIX: Record<string, string> = {
  AC: "ACCOUNTING",
  AM: "ASSESSMENT",
  AR: "ARRANGEMENT",
  AS: "ASSET",
  AT: "ACTOR",
  BP: "PARTNER",
  CL: "COLLATERAL",
  CM: "COMMON",
  CR: "CALCULATOR",
  CU: "CUSTOMER",
  DC: "DOCUMENT",
  DP: "DEPOSIT",
  DT: "DEPARTMENT",
  FX: "FOREIGNEXCHANGE",
  IA: "INTERNALACCOUNT",
  LM: "LIMIT",
  LN: "LOAN",
  PD: "PRODUCT",
  PY: "PAYMENT",
  SF: "STAFF",
  ST: "SETTLEMENT",
  SV: "SERVICEMANAGEMENT",
  TR: "TREASURY",
  UE: "UNDEREXAMINATION",
  XP: "EXTERNALPROXY",
};

export type RulesDomainSelection =
  | { type: "all" }
  | { type: "domain"; domain: string };

export type RulesDomainNavNode = {
  domain: string;
  count: number;
};

export function inferBusinessDomain(serviceCode: string): string {
  const code = (serviceCode || "").trim().toUpperCase();
  if (code.length < 2) return UNCLASSIFIED_DOMAIN;
  return DOMAIN_BY_PREFIX[code.slice(0, 2)] ?? UNCLASSIFIED_DOMAIN;
}

export function domainLabel(domain: string): string {
  if (!domain || domain === UNCLASSIFIED_DOMAIN) return "미분류";
  return domain;
}

export function selectionKey(sel: RulesDomainSelection): string {
  if (sel.type === "all") return "all";
  return `domain:${sel.domain}`;
}

export function matchesDomainSelection(
  item: { businessDomain: string },
  sel: RulesDomainSelection,
): boolean {
  if (sel.type === "all") return true;
  return item.businessDomain === sel.domain;
}

export function buildDomainNavNodes(
  items: { businessDomain: string }[],
): RulesDomainNavNode[] {
  const byDomain = new Map<string, number>();

  for (const item of items) {
    const domain = item.businessDomain || UNCLASSIFIED_DOMAIN;
    byDomain.set(domain, (byDomain.get(domain) ?? 0) + 1);
  }

  return [...byDomain.entries()]
    .map(([domain, count]) => ({ domain, count }))
    .sort((a, b) => {
      if (a.domain === UNCLASSIFIED_DOMAIN) return 1;
      if (b.domain === UNCLASSIFIED_DOMAIN) return -1;
      return a.domain.localeCompare(b.domain, "en");
    });
}

File: backend/app/domain/cbs_service_taxonomy.py
"""CBS service business domain / component taxonomy helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

UNCLASSIFIED_DOMAIN = "UNCLASSIFIED"
UNCLASSIFIED_COMPONENT = ""

# Mirrors the CASE mapping in cbs_srvc.json export SQL.
DOMAIN_BY_PREFIX: dict[str, str] = {
    "AC": "ACCOUNTING",
    "AM": "ASSESSMENT",
    "AR": "ARRANGEMENT",
    "AS": "ASSET",
    "AT": "ACTOR",
    "BP": "PARTNER",
    "CL": "COLLATERAL",
    "CM": "COMMON",
    "CR": "CALCULATOR",
    "CU": "CUSTOMER",
    "DC": "DOCUMENT",
    "DP": "DEPOSIT",
    "DT": "DEPARTMENT",
    "FX": "FOREIGNEXCHANGE",
    "IA": "INTERNALACCOUNT",
    "LM": "LIMIT",
    "LN": "LOAN",
    "PD": "PRODUCT",
    "PY": "PAYMENT",
    "SF": "STAFF",
    "ST": "SETTLEMENT",
    "SV": "SERVICEMANAGEMENT",
    "TR": "TREASURY",
    "UE": "UNDEREXAMINATION",
    "XP": "EXTERNALPROXY",
}


@dataclass(frozen=True, slots=True)
class ServiceTaxonomy:
    business_domain: str
    component_code: str


def infer_business_domain(service_code: str) -> str:
    """Infer domain from the first two characters of SRVC_CD."""
    code = (service_code or "").strip().upper()
    if len(code) < 2:
        return UNCLASSIFIED_DOMAIN
    return DOMAIN_BY_PREFIX.get(code[:2], UNCLASSIFIED_DOMAIN)


def taxonomy_from_raw_row(
    service_code: str,
    raw: Mapping[str, Any] | None,
) -> ServiceTaxonomy:
    """Resolve domain/component from a cbs_srvc raw row, with prefix fallback."""
    domain = ""
    component = ""
    if raw:
        domain = str(
            raw.get("business_domain")
            or raw.get("BUSINESS_DOMAIN")
            or ""
        ).strip()
        component = str(
            raw.get("component_code")
            or raw.get("CMPNT_CD")
            or ""
        ).strip()
    if not domain:
        domain = infer_business_domain(service_code)
    return ServiceTaxonomy(
        business_domain=domain or UNCLASSIFIED_DOMAIN,
        component_code=component,
    )


def build_taxonomy_map(
    raw_by_code: Mapping[str, Mapping[str, Any]],
) -> dict[str, ServiceTaxonomy]:
    """Map service_code → taxonomy for every known catalog row."""
    out: dict[str, ServiceTaxonomy] = {}
    for code, row in raw_by_code.items():
        key = (code or "").strip()
        if not key:
            continue
        out[key] = taxonomy_from_raw_row(key, row)
    return out
